From f4b5ee5ef04627f42ab9b4c01172d1aa76c5a67d Mon Sep 17 00:00:00 2001
From: Liu Heng <liuhenga@uniontech.com>
Date: Mon, 16 Dec 2024 20:49:32 +0800
Subject: [PATCH] Give the popup window focus

Since the popup window gets events through forwardToPopup and
is not in an active state itself, the focus is not on the popup
window, which causes the popup window to be unable to receive
any input from the input method. When there is an activePopupWindow,
it is prioritized to consider the popup window as focusWindow.
This achieves a similar effect to forwardFocusToPopup, and upon
activating the popup, the focus is actively set to the
corresponding popup window, allowing the popup window to
properly receive input from the input method.

Fixes: QTBUG-132250
Pick-to: 6.8 6.9
Change-Id: I83dbf3bc089f6a7197e8c03cdb7289e3b8e873a7
---
 src/gui/kernel/qguiapplication.cpp | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/src/gui/kernel/qguiapplication.cpp b/src/gui/kernel/qguiapplication.cpp
index dcdff702c4..60b3baac6f 100644
--- a/src/gui/kernel/qguiapplication.cpp
+++ b/src/gui/kernel/qguiapplication.cpp
@@ -998,12 +998,22 @@ void QGuiApplicationPrivate::activatePopup(QWindow *popup)
     popup_list.removeOne(popup); // ensure that there's only one entry, and it's the last
     qCDebug(lcPopup) << "appending popup" << popup << "to existing" << popup_list;
     popup_list.append(popup);
+
+    // Make the Popup window have focus so that focus-based event distribution, such as input
+    // method input, can be handled correctly
+    QWindowSystemInterface::handleFocusWindowChanged(popup);
 }
 
 bool QGuiApplicationPrivate::closePopup(QWindow *popup)
 {
     const auto removed = QGuiApplicationPrivate::popup_list.removeAll(popup);
     qCDebug(lcPopup) << "removed?" << removed << "popup" << popup << "; remaining" << popup_list;
+
+    if (removed) {
+        // Return the focus to the topmost popup window or the last window
+        QWindowSystemInterface::handleFocusWindowChanged(QGuiApplicationPrivate::focus_window);
+    }
+
     return removed; // >= 1 if something was removed
 }
 
@@ -1029,6 +1039,11 @@ bool QGuiApplicationPrivate::closeAllPopups()
 */
 QWindow *QGuiApplication::focusWindow()
 {
+    // Since keyboard and mouse events are forwarded to the popup window,
+    // the focus also needs to be assigned to the popup window.
+    QWindow *window = QGuiApplicationPrivate::activePopupWindow();
+    if (window)
+        return window;
     return QGuiApplicationPrivate::focus_window;
 }
 
-- 
2.45.2

